src/api.ts

```diff
@@ -0,0 +1,395 @@
+ *
+ * Only `application/json` is supported.
+ *
+ * TODO: provide example usage after api is stabilized
+ *
+ * @see https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-api-integration-types.html
+ */
+export class MockApiIntegration<
+  Props extends MockApiIntegrationProps<any, any, any>
+> implements BaseApiIntegration
+{
+  /**
+   * This static property identifies this class as a MockApiIntegration to the Functionless plugin.
+   */
+  public static readonly FunctionlessType = "MockApiIntegration";
+
+  private readonly request: FunctionDecl;
+  private readonly responses: { [K in keyof Props["responses"]]: FunctionDecl };
+
+  public constructor(props: Props) {
+    this.request = validateFunctionDecl(props.request);
+    this.responses = Object.fromEntries(
+      Object.entries(props.responses).map(([k, v]) => [
+        k,
+        validateFunctionDecl(v),
+      ])
+    ) as { [K in keyof Props["responses"]]: FunctionDecl };
+  }
+
+  addMethod(httpMethod: HttpMethod, resource: aws_apigateway.Resource): void {
+    const requestTemplate = toVTL(this.request, "request");
+
+    const responseEntries: [string, FunctionDecl][] = Object.entries(
+      this.responses
+    );
+    const integrationResponses: aws_apigateway.IntegrationResponse[] =
+      responseEntries.map(([statusCode, fn]) => ({
+        statusCode,
+        responseTemplates: {
+          "application/json": toVTL(fn, "response"),
+        },
+        selectionPattern: `^${statusCode}$`,
+      }));
+
+    const integration = new aws_apigateway.MockIntegration({
+      requestTemplates: {
+        "application/json": requestTemplate,
+      },
+      integrationResponses,
+    });
+
+    const methodResponses = Object.keys(this.responses).map((statusCode) => ({
+      statusCode,
+    }));
+
+    // TODO: support requestParameters, authorizers, models and validators
+    resource.addMethod(httpMethod, integration, {
+      methodResponses,
+    });
+  }
+}
+
+export interface AwsApiIntegrationProps<
+  Request extends ApiRequestProps<any, any, any, any>,
+  IntegrationRequest,
+  IntegrationResponse,
+  MethodResponse
+> {
+  /**
+   * Map API request to an integration request.
+   */
+  request: RequestTransformerFunction<Request, IntegrationRequest>;
+  /**
+   * Integration target backing this API. The result of `request` will be sent.
+   */
+  integration: IntegrationTarget<IntegrationRequest, IntegrationResponse>;
+  /**
+   * Map integration response to a method response.
+   * TODO: we need to handle multiple responses
+   */
+  response: ResponseTransformerFunction<IntegrationResponse, MethodResponse>;
+}
+
+/**
+ * An AWS API Gateway integration lets you integrate an API with an AWS service
+ * supported by Functionless. The request is transformed via VTL and sent to the
+ * service via API call, and the response is transformed via VTL and returned in
+ * the response.
+ *
+ * TODO: we need to support multiple responses
```

Comment:
Are we tracking this in an issue or is it part of this change?

Reply:
it'll be part of this change